### ISSUE TITLE
Status bar showing idle status for a busy kernel when switching notebooks

### ISSUE THREAD

**Issue by kiranik** [jupyterlab/jupyterlab#14578]
## Description

Jupyter lab status bar is showing that the kernel is idle even though the kernel is busy as seen in the top right kernel status tab. This occurs when we switch between an idle notebook and a busy notebook.

## Reproduce

1. Create a new notebook named **notebook_1.ipynb**. Add a long running script in a cell and execute the same.
2. Create a second notebook named **notebook_2.ipynb**. Assign a kernel and wait until kernel is idle.
3. Switch between notebook_1.ipynb and notebook_2.ipynb.
4. Check status bar of notebook_1.ipynb. Kernel status will be shown as idle instead of busy.

## Expected behaviour

Kernel status of busy notebook should be shown as _busy_ instead of _idle_.

## Context

- Operating System and version: macOS Monterey, version 12.6.4
- Browser and version: Chrome 113.0.5672.92 
- JupyterLab version: 3.6.3 and 4.0.0

<details><summary>package versions</summary>
<pre>
aiofiles==22.1.0
aiosqlite==0.19.0
anyio==3.6.2
appnope==0.1.3
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
arrow==1.2.3
attrs==23.1.0
Babel==2.12.1
backcall==0.2.0
beautifulsoup4==4.12.2
bleach==6.0.0
cached-property==1.5.2
certifi==2023.5.7
cffi==1.15.1
charset-normalizer==3.1.0
debugpy==1.6.7
decorator==5.1.1
defusedxml==0.7.1
entrypoints==0.4
fastjsonschema==2.17.1
fqdn==1.5.1
idna==3.4
importlib-metadata==6.6.0
importlib-resources==5.12.0
ipykernel==6.16.2
ipython==7.34.0
ipython-genutils==0.2.0
isoduration==20.11.0
jedi==0.18.2
Jinja2==3.1.2
json5==0.9.14
jsonpointer==2.3
jsonschema==4.17.3
jupyter-events==0.6.3
jupyter-server==1.24.0
jupyter-ydoc==0.2.4
jupyter_client==7.4.9
jupyter_core==4.12.0
jupyter_server_fileid==0.9.0
jupyter_server_ydoc==0.8.0
jupyterlab==3.6.3
jupyterlab-pygments==0.2.2
jupyterlab_server==2.22.1
MarkupSafe==2.1.2
matplotlib-inline==0.1.6
mistune==2.0.5
nbclassic==1.0.0
nbclient==0.7.4
nbconvert==7.4.0
nbformat==5.8.0
nest-asyncio==1.5.6
notebook==6.5.4
notebook_shim==0.2.3
packaging==23.1
pandocfilters==1.5.0
parso==0.8.3
pexpect==4.8.0
pickleshare==0.7.5
pkgutil_resolve_name==1.3.10
prometheus-client==0.16.0
prompt-toolkit==3.0.38
psutil==5.9.5
ptyprocess==0.7.0
pycparser==2.21
Pygments==2.15.1
pyrsistent==0.19.3
python-dateutil==2.8.2
python-json-logger==2.0.7
pytz==2023.3
PyYAML==6.0
pyzmq==25.0.2
requests==2.30.0
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
Send2Trash==1.8.2
six==1.16.0
sniffio==1.3.0
soupsieve==2.4.1
terminado==0.17.1
tinycss2==1.2.1
tomli==2.0.1
tornado==6.2
traitlets==5.9.0
typing_extensions==4.5.0
uri-template==1.2.0
urllib3==2.0.2
wcwidth==0.2.6
webcolors==1.13
webencodings==0.5.1
websocket-client==1.5.2
y-py==0.5.9
ypy-websocket==0.8.2
zipp==3.15.0
</pre>
</details>

https://github.com/jupyterlab/jupyterlab/assets/6435143/46594eec-75e3-4fcb-9ece-4b25360dedbd

**Comment by brijsiyag** [comment]
This problem is not in Jupyterlab. Actually IOpub message is received twice with alternate `execution_state` message as `idle` and `busy` , logged as below.
```
future.ts:260 busy
future.ts:260 idle
future.ts:260 busy
future.ts:260 idle
future.ts:260 busy
future.ts:260 idle
future.ts:260 busy
future.ts:260 idle
future.ts:260 busy
future.ts:260 idle
```
I think we should report this issue to https://github.com/ipython/ipykernel repo.

**Comment by FoSuCloud** [comment]
> 
You are right. I observed from websocket that switching the ipynb file will send a debug_request message, then the kernel responds to an iopub message, execution_state is busy, and then continues to respond to an iopub message, execution_state is idle

**Comment by kiranik** [comment]
Opened this issue as [ipython/ipykernel#1121](https://github.com/ipython/ipykernel/issues/1121).

**Comment by ojarjur** [comment]
FWIW, I ran into this issue too, investigated it, and came to a very similar conclusion as @Zsailer in https://github.com/jupyterlab/jupyterlab/issues/14578#issuecomment-1593236064

That being said, there *also* appears to be the same issue in jupyter-server, as [every status message updates the kernel's  `execution_state` field regardless of parent message](https://github.com/jupyter-server/jupyter_server/blob/95798620a5b3d8b69e7e03f895c4e2bf7227677c/jupyter_server/services/kernels/kernelmanager.py#L562)

What seems more concerning to me, however, is that we seem to have logic for tracking the kernel status in at least 3 different places; in the jupyter-server package for determining the kernel's `execution_state` field, and in at least two different places within JupyterLab (for the upper-right-hand-corner indicator and in the status bar).

I think the best fix would be to correct the behavior in `jupyter-server`, and then have the JupyterLab UI use the value of the kernel's `execution_state` field. I'm reasonably certain that we are not currently doing that, as I changed the behavior of `jupyter-server` on my machine to fix this and still observed the inconsistent behavior in the UI reported in this issue.

**Comment by Lewic1201** [comment]
Can this problem be solved or not?

**Comment by Lewic1201** [comment]
> FWIW, I ran into this issue too, investigated it, and came to a very similar conclusion as @Zsailer in [#14578 (comment)](https://github.com/jupyterlab/jupyterlab/issues/14578#issuecomment-1593236064)
> 
> That being said, there _also_ appears to be the same issue in jupyter-server, as [every status message updates the kernel's `execution_state` field regardless of parent message](https://github.com/jupyter-server/jupyter_server/blob/95798620a5b3d8b69e7e03f895c4e2bf7227677c/jupyter_server/services/kernels/kernelmanager.py#L562)
> 
> What seems more concerning to me, however, is that we seem to have logic for tracking the kernel status in at least 3 different places; in the jupyter-server package for determining the kernel's `execution_state` field, and in at least two different places within JupyterLab (for the upper-right-hand-corner indicator and in the status bar).
> 
> I think the best fix would be to correct the behavior in `jupyter-server`, and then have the JupyterLab UI use the value of the kernel's `execution_state` field. I'm reasonably certain that we are not currently doing that, as I changed the behavior of `jupyter-server` on my machine to fix this and still observed the inconsistent behavior in the UI reported in this issue.

Can you share how you solved it?

**Comment by ojarjur** [comment]
> > FWIW, I ran into this issue too, investigated it, and came to a very similar conclusion as @Zsailer in [#14578 (comment)](https://github.com/jupyterlab/jupyterlab/issues/14578#issuecomment-1593236064)
> > That being said, there _also_ appears to be the same issue in jupyter-server, as [every status message updates the kernel's `execution_state` field regardless of parent message](https://github.com/jupyter-server/jupyter_server/blob/95798620a5b3d8b69e7e03f895c4e2bf7227677c/jupyter_server/services/kernels/kernelmanager.py#L562)
> > What seems more concerning to me, however, is that we seem to have logic for tracking the kernel status in at least 3 different places; in the jupyter-server package for determining the kernel's `execution_state` field, and in at least two different places within JupyterLab (for the upper-right-hand-corner indicator and in the status bar).
> > I think the best fix would be to correct the behavior in `jupyter-server`, and then have the JupyterLab UI use the value of the kernel's `execution_state` field. I'm reasonably certain that we are not currently doing that, as I changed the behavior of `jupyter-server` on my machine to fix this and still observed the inconsistent behavior in the UI reported in this issue.
> 
> Can you share how you solved it?

[This](https://github.com/ojarjur/jupyter_server/commit/30b45dbace700381203b5a07f3c8e0335fce0271) is my proposed change to fix the issue on the server side.

I didn't send out a PR yet because I wanted to solicit some more feedback on the general approach before doing that.

**Comment by krassowski** [comment]
> [This](https://github.com/ojarjur/jupyter_server/commit/30b45dbace700381203b5a07f3c8e0335fce0271) is my proposed change to fix the issue on the server side.

That ended up merged as https://github.com/jupyter-server/jupyter_server/pull/1429 later on, right?

So `jupyter-server` [v2.15.0](https://github.com/jupyter-server/jupyter_server/releases/tag/v2.15.0) should no longer exhibit the issue?

**Comment by krassowski** [comment]
This is the status bar implementation:

https://github.com/jupyterlab/jupyterlab/blob/11a9a7c46cffa2b63fd53a12ce0d2bbf1286590d/packages/apputils/src/kernelstatuses.tsx#L199-L264

And this is the toolbar implementation:

https://github.com/jupyterlab/jupyterlab/blob/11a9a7c46cffa2b63fd53a12ce0d2bbf1286590d/packages/notebook/src/executionindicator.tsx#L298-L398

The toolbar stores a copy of the status information on per notebook basis, whereas status bar pulls the information from `sessionContext`.

So both use [`kernelDisplayStatus`](https://github.com/jupyterlab/jupyterlab/blob/main/packages/apputils/src/sessioncontext.tsx#L595-L628) (which conflates status from different kernel channels)...

**Comment by krassowski** [comment]
@Zsailer you mention that this is because frontend conflates the `idle` messages from different channels. Do you happen to remember how this happens? Looking at the code I see only one usage of `‎KernelConnection._updateStatus` that could possibly set the `idle` value, but that one does have a guard on `iopub` channel:

<img width="471" height="471" alt="Image" src="https://github.com/user-attachments/assets/ad139e6e-9c5a-40e4-afb8-ecca4510f418" />

https://github.com/jupyterlab/jupyterlab/blob/11a9a7c46cffa2b63fd53a12ce0d2bbf1286590d/packages/services/src/kernel/default.ts#L1630-L1655

**Comment by krassowski** [comment]
Testing with `jupyter-server` `2.17.0` to me this _seems_ as if the issue was still at the server side:

<img width="3411" height="2049" alt="Image" src="https://github.com/user-attachments/assets/67a67e6e-99c0-4e46-b9ce-6a3a36abf3eb" />

I'm happy to be proven wrong, just don't see where the issue at the frontend side is!

**Comment by krassowski** [comment]
Looking at the kernel spy, it might be that it is neither server nor frontends fault but kernels? When switching tabs JupyterLab sends `debug_request` in the `control` channel but this generates extra `status` messages on the `iopub` channel.

<img width="818" height="1601" alt="Image" src="https://github.com/user-attachments/assets/4f935f88-453c-43e5-96ca-4ab9fe9db85b" />

Another issue is that we send two `debug_request`s probably needlessly.

**Comment by krassowski** [comment]
Ah, now I see what Zach was saying here:

> The indicator should only represent the status messages (coming from IOPub) that were trigged by parent messages on shell channel, since the shell channel represents the kernel's execution state; however, JupyterLab is just showing the last status message, regardless of the channel. It isn't filtering for shell messages. The "idle" you're seeing here is likely from the control channel.

Basically that we should filter on parent message channel, not only on the status message channel.

I think this would warrant a clarification on https://jupyter-client.readthedocs.io/en/stable/messaging.html#kernel-status - it does not describe this behaviour currently.

**Comment by Zsailer** [comment]
Yes, for reference, see these issues I opened in Jupyter client a few years ago:

https://github.com/jupyter/jupyter_client/issues/838

https://github.com/jupyter/jupyter_client/issues/839

**Comment by Zsailer** [comment]
I've gotten around this issue in a custom kernel manager by keeping a message cache in the server that tracks what channel each status message came from.

**Comment by krassowski** [comment]
Reopening as the PR https://github.com/jupyterlab/jupyterlab/pull/17894 was reverted to fix CI. I hope we can merge it again soon.
